File: instructor Henry/lenguajes/Java/clases abstractas/clase abstracta.java
// Clase abstracta en java

/*En Java, se usa la palabra clave abstract para definir tanto clases como métodos abstractos. Una clase que 
contiene al menos un método abstracto debe ser abstracta.
*/
abstract class Animal {
    abstract void emitirSonido(); // Método abstracto

    void dormir() {
        System.out.println("Zzz");
    }
}

class Dog extends Animal {
    void emitirSonido() {
        System.out.println("Woof!");
    }
}

public class Main {
    public static void main(String[] args) {
        Dog dog = new Dog();
        dog.emitirSonido(); // Woof!
        dog.dormir();     // Zzz
    }
}

File: instructor Henry/lenguajes/Python/clases abstractas/clase abstracta.py
# Clases en Python

#En Python, las clases abstractas se definen utilizando el módulo abc (Abstract Base Classes). 
#Se usa el decorador @abstractmethod para definir métodos abstractos.

from abc import ABC, abstractmethod

class Animal(ABC):
    @abstractmethod
    def emitir_sonido(self):
        pass

    def dormir(self):
        print("Zzz")

class Dog(Animal):
    def emitir_sonido(self):
        print("Woof!")

dog = Dog()
dog.emitir_sonido()  # Woof!
dog.dormir()       # Zzz
